aten/src/ATen/native/Pad.cpp

```diff
@@ -0,0 +1,271 @@
+              other_width_before + self.size(slice_dim_idx),
+              at::indexing::None)));
+        }
+      }
+      result.index_put_(slices, constant_values[pad_dim_idx][0]);
+    }
+
+    // Fill after-padding for this dimension
+    if (width_after > 0) {
+      std::vector<at::indexing::TensorIndex> slices;
+      for (int64_t slice_dim_idx = 0; slice_dim_idx < self.dim(); slice_dim_idx++) {
+        if (slice_dim_idx == pad_dim_idx) {
+          slices.push_back(at::indexing::TensorIndex(at::indexing::Slice(
+              width_before + self.size(slice_dim_idx),
+              result.size(slice_dim_idx),
+              at::indexing::None)));
+        } else if (slice_dim_idx < pad_dim_idx) {
+          slices.push_back(at::indexing::TensorIndex(at::indexing::Slice(
+              0,
+              result.size(slice_dim_idx),
+              at::indexing::None)));
+        } else {
+          auto other_width_before = pad_width[slice_dim_idx][0].item<int64_t>();
+          slices.push_back(at::indexing::TensorIndex(at::indexing::Slice(
+              other_width_before,
+              other_width_before + self.size(slice_dim_idx),
+              at::indexing::None)));
+        }
+      }
+      result.index_put_(slices, constant_values[pad_dim_idx][1]);
+    }
+  }
+}
+
+void check_pad_specifier_is_none(const c10::optional<Tensor>& arg, const char* arg_name, c10::string_view mode_str) {
+  TORCH_CHECK(!arg.has_value(),
+    "Unsupported keyword argument for '", std::string(mode_str), "' mode: ", arg_name);
+}
+
+// Calculate the size of the padded tensor
+//
+// NOTE: pad_width must be processed with expand_pad_specifier_arg before
+// calling this function
+Tensor get_result_size(const Tensor& self, const Tensor& pad_width) {
+  auto result_size_tensor = at::tensor(
+    self.sizes(),
+    TensorOptions().dtype(at::kLong).device(at::kCPU).memory_format(c10::MemoryFormat::Contiguous));
+
+  result_size_tensor += pad_width.sum(1).cpu();
+
+  return result_size_tensor;
+}
+
+// Create an IntArrayRef from a tensor
+IntArrayRef tensor_to_arrayref(const Tensor& size_tensor) {
+  return IntArrayRef(size_tensor.data_ptr<int64_t>(), size_tensor.numel());
+}
+
+// Check that a tensor is on the expected device
+void check_device(const Tensor& arg, const char* arg_name, at::Device self_device) {
+  TORCH_CHECK(arg.device() == self_device,
+    "Expected '", arg_name, "' to be on the same device as 'self' (",
+    self_device, ") but got ", arg.device());
+}
+
+Tensor& pad_out_impl(
+  const Tensor& self,
+  const Tensor& pad_width,
+  c10::string_view mode_str,
+  const c10::optional<Tensor>& constant_values_opt,
+  Tensor& result
+) {
+  TORCH_CHECK(pad_width.device() == at::kCPU,
+    "Expected 'pad_width' to be on CPU, but got ", pad_width.device());
+  PadMode mode = get_pad_mode_from_str(mode_str);
+
+  Tensor pad_width_ = pad_width;
+  // TODO: It would probably be better to properly handle these types of
```

Comment:
Would it be more natural to assert pad_width is long and require C++ API users to specify its dtype correctly? Is there a chance that the Python API would produce a non-long tensor for this?

Reply:
I've disallowed all types for `pad_width` except `torch.long`